doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -471,57 +437,26 @@ stands for :n:`let @ident := fun {+ @binder} => @term in @term’`.
 Definition by cases: match
 --------------------------
 
-.. insertgram term_match record_pattern
-
-.. productionlist:: coq
-   term_match : match `case_items_comma` `return_type_opt` with `or_opt` `eqns_or_opt` end
-   case_items_comma : `case_items_comma` , `case_item`
-                    : `case_item`
-   return_type_opt : `return_type`
-                   : `empty`
-   as_return_type_opt : `as_name_opt` `return_type`
-                      : `empty`
-   return_type : return `term100`
-   case_item : `term100` `as_name_opt` `in_opt`
-   as_name_opt : as `name`
-               : `empty`
-   in_opt : in `pattern`
-          : `empty`
-   or_opt : |
-          : `empty`
-   eqns_or_opt : `eqns_or`
-               : `empty`
-   eqns_or : `eqns_or` | `eqn`
-           : `eqn`
-   eqn : `patterns_comma_list_or` => `term`
-   patterns_comma_list_or : `patterns_comma_list_or` | `patterns_comma`
-                          : `patterns_comma`
-   patterns_comma : `patterns_comma` , `pattern`
-                  : `pattern`
-   pattern : `pattern10` : `term`
-           : `pattern10`
-   pattern10 : `pattern1` as `name`
-             : `pattern1_list`
-             : @ `qualid` `pattern1_list_opt`
-             : `pattern1`
-   pattern1_list : `pattern1_list` `pattern1`
-                 : `pattern1`
-   pattern1_list_opt : `pattern1_list`
-                     : `empty`
-   pattern1 : `pattern0` % `ident`
-            : `pattern0`
-   pattern0 : `qualid`
-            : {| `record_patterns_opt` |}
-            : _
-            : ( `patterns_or` )
-            : `numeral`
-            : `string`
-   patterns_or : `patterns_or` | `pattern`
-               : `pattern`
-   record_patterns_opt : `record_patterns_opt` ; `record_pattern`
-                       : `record_pattern`
-                       : `empty`
-   record_pattern : `qualid` := `pattern`
+.. insertprodn term_match pattern0
+
+.. prodn::
+   term_match ::= match {+, @case_item } {? return @term100 } with {? %| } {*| @eqn } end
+   case_item ::= @term100 {? as @name } {? in @pattern }
+   eqn ::= {+| {+, @pattern } } => @term
+   pattern ::= @pattern10 : @term
+   | @pattern10
+   pattern10 ::= @pattern1 as @name
+   | {+ @pattern1 }
+   | @ @qualid {* @pattern1 }
```

Comment:
Because that paragraph (and maybe the code) is incorrect?  There are only 2 productions in `pattern` that include `qualid` (renamed from `Prim.reference`):

```
  pattern:
    | "10" LEFTA
      | "@"; r = Prim.reference; lp = LIST0 NEXT ->
        { CAst.make ~loc @@ CPatCstr (r, Some lp, []) } ]
    | "0"
      [ r = Prim.reference                -> { CAst.make ~loc @@ CPatAtom (Some r) }
```

What should replace `@qualid {* @ident}` in the paragraph?

Reply:
Actually, the paragraph is correct, as it describes the simplest form of pattern matching, e.g. `Definition foo n := match n with O => true | S n => false end.` However, I didn't realize but the grammar as presented already allows to build such a construction thanks to the `{+ @pattern1 }` clause (`pattern1` can just be refined into `qualid`). So there is nothing to fix here.

However, I notice in passing that `{+ @pattern1 }` and `@pattern1` are redundant. The second one could be removed in favor of the former. And for more symmetry with the form `@qualid {* @pattern1 }`, `{+ @pattern1 }` could be displayed as `@pattern1 {* @pattern1 }`.

Reply:
> However, I notice in passing that {+ @pattern1 } and @pattern1 are redundant.

I noticed that, but in this case I thought maybe it's worth keeping the pattern -> pattern10 -> pattern1 -> pattern0 chain very explicit.  I will change it, though.

> And for more symmetry with the form `@ @qualid {* @pattern1 }`, `{+ @pattern1 }` could be displayed as `@pattern1 {* @pattern1 }`.

I assume you meant `@ @pattern1 {* @pattern1 }`, right?  Wouldn't that incorrectly include strings in the form `@ @pattern0 % @ident`?

Reply:
> > 
> 
> I assume you meant `@ @pattern1 {* @pattern1 }`, right? Wouldn't that incorrectly include strings in the form `@ @pattern0 % @ident`?

I don't understand. No leading `@` of course, otherwise that would not be the same grammar. I don't understand your question either.

Reply:
Got it.  My bad.